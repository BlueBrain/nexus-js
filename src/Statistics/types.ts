export interface StatisticsResponse {
  '@context': string;
  totalEvents: number;
  processedEvents: string;
  remainingEvents: string;
  evaluatedEvents: string;
  discardedEvents: string;
  lastEventDateTime: string;
  lastProcessedEventDateTime: string;
  delayInSeconds: string;
}
