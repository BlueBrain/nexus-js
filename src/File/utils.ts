// @ts-ignore
import FormData from 'isomorphic-form-data';
import { NexusFileResponse } from './types';
import { httpPost, HttpConfigTypes, httpGet } from '../utils/http';
import NexusFile from './index';
import { isBrowser } from '../utils';
import { ReadStream } from 'fs';
import { Readable } from 'stream';

export async function createFile(
  orgLabel: string,
  projectLabel: string,
  file: File | Blob | ReadableStream | ReadStream | Readable,
): Promise<NexusFile> {
  try {
    // create new form data
    const form = new FormData();
    // add file
    form.append('file', file);
    // if in Node.js, we need to manually set headers
    let extraHeaders = {};
    if (!isBrowser) {
      extraHeaders = form.getHeaders();
    }
    const fileResponse: NexusFileResponse = await httpPost(
      `/files/${orgLabel}/${projectLabel}`,
      form,
      {
        extraHeaders,
        sendAs: HttpConfigTypes.FILE,
      },
    );
    return new NexusFile(orgLabel, projectLabel, fileResponse);
  } catch (error) {
    throw error;
  }
}

export async function getFileSelf(
  selfUrl: string,
  shouldFetchFile: boolean = false,
): Promise<NexusFile> {
  const fileResponse: NexusFileResponse = await httpGet(selfUrl, {
    useBase: false,
  });
  const [id, schema, projectLabel, orgLabel, ...rest] = selfUrl
    .split('/')
    .reverse();
  const file = new NexusFile(orgLabel, projectLabel, fileResponse);
  if (shouldFetchFile) {
    await file.getFile();
  }
  return file;
}

export async function getFile(
  orgLabel: string,
  projectLabel: string,
  fileId: string,
  shouldFetchFile: boolean = false,
): Promise<NexusFile> {
  const fileResponse: NexusFileResponse = await httpGet(
    `/files/${orgLabel}/${projectLabel}/${fileId}`,
  );
  const file = new NexusFile(orgLabel, projectLabel, fileResponse);
  if (shouldFetchFile) {
    await file.getFile();
  }
  return file;
}

export async function getRawFile(
  selfURL: string,
): Promise<string | Blob | ReadStream | ArrayBuffer> {
  const rawFile: string | Blob | ReadStream | ArrayBuffer = await httpGet(
    selfURL,
    {
      useBase: false,
      extraHeaders: {
        Accept: '*/*',
      },
      receiveAs: HttpConfigTypes.BASE64,
    },
  );
  return rawFile;
}
