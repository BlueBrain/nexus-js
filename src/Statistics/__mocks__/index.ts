export const mockGetStatistics = jest.fn();

const mockStatisticsClass = jest.fn().mockImplementation(() => ({
  getStatistics: mockGetStatistics,
}));

export default mockStatisticsClass;
