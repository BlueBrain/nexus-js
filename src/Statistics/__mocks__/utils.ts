export const getViewStatistics = jest.fn();
export const getResourceViewStatistics = jest.fn();
