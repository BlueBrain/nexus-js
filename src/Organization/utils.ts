import Organization from '.';
import {
  ListOrgOptions,
  CreateOrgPayload,
  OrgResponse,
  ListOrgResponse,
  OrgResponseCommon,
  OrgCreatedEvent,
  OrgUpdatedEvent,
  OrgDeprecatedEvent,
  OrgEventListeners,
  OrgEventType,
} from './types';
import { httpPut, httpGet, httpDelete } from '../utils/http';
import { CreateOrganizationException } from './exceptions';
import { PaginatedList } from '../utils/types';
import { getEventSource, parseMessageEventData } from '../utils/events';
// @ts-ignore
import EventSource = require('eventsource');
import { buildQueryParams } from '../utils';

/**
 *
 * @param label The label of your organization
 * @param name The name of your organization
 */
export async function createOrganization(
  label: string,
  orgPayload?: CreateOrgPayload,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(
      `/orgs/${label}`,
      orgPayload,
    );
    return new Organization({ ...orgResponse, ...orgPayload });
  } catch (error) {
    throw new CreateOrganizationException(error.message);
  }
}

/**
 *
 * @param orgLabel The organization label to fetch
 * @param Object<revision, tag> The specific tag OR revision to fetch
 */
export async function getOrganization(
  orgLabel: string,
  options?: { rev?: number; tag?: string },
): Promise<Organization> {
  try {
    const opts = buildQueryParams(options);
    const orgResponse: OrgResponse = await httpGet(`/orgs/${orgLabel}${opts}`);
    const org = new Organization(orgResponse);
    return org;
  } catch (error) {
    throw new Error(error.message);
  }
}

export async function listOrganizations(
  options?: ListOrgOptions,
): Promise<PaginatedList<Organization>> {
  const opts = buildQueryParams(options);
  try {
    const listOrgResponse: ListOrgResponse = await httpGet(`/orgs${opts}`);
    if (listOrgResponse.code || !listOrgResponse._results) {
      return {
        total: 0,
        index: 0,
        results: [],
      };
    }

    const orgs: Organization[] = listOrgResponse._results.map(
      (commonResponse: OrgResponseCommon) =>
        new Organization({
          ...commonResponse,
          '@context': listOrgResponse['@context'],
        }),
    );

    return {
      total: listOrgResponse._total,
      index: (options && options.from) || 1,
      results: orgs,
    };
  } catch (error) {
    throw error;
  }
}

/**
 *
 * @param orgLabel Current organization label
 * @param rev Last known revision
 * @param newName Knew name the organization will be called with
 */
export async function updateOrganization(
  orgLabel: string,
  rev: number,
  orgPayload: CreateOrgPayload,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(
      `/orgs/${orgLabel}?rev=${rev}`,
      orgPayload,
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw error;
  }
}

/**
 *
 * @param orgLabel Label of the Org to be deprecated
 * @param rev Last know revision
 */
export async function deprecateOrganization(
  orgLabel: string,
  rev: number,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpDelete(
      `/orgs/${orgLabel}?rev=${rev}`,
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw error;
  }
}

export function subscribe(listeners: OrgEventListeners): EventSource {
  const event: EventSource = getEventSource('/orgs/events');

  // set event listeners
  listeners.onOpen && (event.onopen = listeners.onOpen);
  listeners.onError && (event.onerror = listeners.onError);
  listeners.onOrgCreated &&
    event.addEventListener(OrgEventType.OrgCreated, (event: Event) =>
      parseMessageEventData<OrgCreatedEvent>(event as MessageEvent)(
        listeners.onOrgCreated,
      ),
    );
  listeners.onOrgUpdated &&
    event.addEventListener(OrgEventType.OrgUpdated, (event: Event) =>
      parseMessageEventData<OrgUpdatedEvent>(event as MessageEvent)(
        listeners.onOrgUpdated,
      ),
    );
  listeners.onOrgDeprecated &&
    event.addEventListener(OrgEventType.OrgDeprecated, (event: Event) =>
      parseMessageEventData<OrgDeprecatedEvent>(event as MessageEvent)(
        listeners.onOrgDeprecated,
      ),
    );

  return event;
}
