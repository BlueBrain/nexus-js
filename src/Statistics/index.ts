import { StatisticsResponse } from './types';
import { getViewStatistics, getResourceStatistics } from './utils';

export default class Statistics {
  static getForView = getViewStatistics;
  static getForResource = getResourceStatistics;

  readonly context: string;
  readonly totalEvents: number;
  readonly processedEvents: string;
  readonly remainingEvents: string;
  readonly evaluatedEvents: string;
  readonly discardedEvents: string;
  readonly lastEventDateTime: string;
  readonly lastProcessedEventDateTime: string;
  readonly delayInSeconds: string;

  constructor(response: StatisticsResponse) {
    this.context = response['@context'];
    this.totalEvents = response.totalEvents;
    this.processedEvents = response.processedEvents;
    this.remainingEvents = response.remainingEvents;
    this.evaluatedEvents = response.evaluatedEvents;
    this.discardedEvents = response.discardedEvents;
    this.lastEventDateTime = response.lastEventDateTime;
    this.lastProcessedEventDateTime = response.lastProcessedEventDateTime;
    this.delayInSeconds = response.delayInSeconds;
  }
}
