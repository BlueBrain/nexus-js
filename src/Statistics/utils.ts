import { httpGet } from '../utils/http';
import { StatisticsResponse } from './types';
import Statistics from '.';

export async function getViewStatistics(
  orgLabel: string,
  projectLabel: string,
  viewId: string,
): Promise<Statistics> {
  try {
    const data: StatisticsResponse = await httpGet(
      `/views/${orgLabel}/${projectLabel}/${encodeURIComponent(
        viewId,
      )}/statistics`,
    );
    return new Statistics(data);
  } catch (error) {
    throw error;
  }
}

export async function getResourceStatistics(
  orgLabel: string,
  projectLabel: string,
  schemaId: string,
  viewId: string,
): Promise<Statistics> {
  try {
    const data: StatisticsResponse = await httpGet(
      `/resources/${orgLabel}/${projectLabel}/${encodeURIComponent(
        schemaId,
      )}/${encodeURIComponent(viewId)}/statistics`,
    );
    return new Statistics(data);
  } catch (error) {
    throw error;
  }
}
