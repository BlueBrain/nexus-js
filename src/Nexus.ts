import store from './store';
import Organization from './Organization';
import { CreateOrgPayload, ListOrgOptions } from './Organization/types';
import makeOrgUtils, { OrgUtils } from './Organization/utils';
import { PaginatedList } from './utils/types';
import Store from './utils/Store';

type NexusConfig = {
  environment?: string;
  token?: string;
};

// Without this, jest crashes really bad...
if (process.env.NODE === 'TEST') {
  Organization;
}

export default class Nexus {
  readonly store: Store;
  readonly Organization: OrgUtils;

  // update global store
  static setEnvironment(environment: string): void {
    store.update('api', state => ({
      ...state,
      baseUrl: environment,
    }));
  }

  // update global store
  static setToken(token: string): void {
    if (!token || token === undefined || token.length === 0) {
      throw new Error('Token is invalid.');
    }
    store.update('auth', state => ({
      ...state,
      accessToken: token,
    }));
  }

  // update global store
  static removeToken(): void {
    store.update('auth', state => ({
      ...state,
      accessToken: undefined,
    }));
  }

  constructor(config: NexusConfig) {
    if (!config.environment) {
      throw new Error('No environment provided');
    }
    this.store = new Store({
      auth: { accessToken: config.token },
      api: { baseUrl: config.environment },
    });

    this.Organization = makeOrgUtils(this.store);
  }

  setToken(token: string) {
    if (!token || token === undefined || token.length === 0) {
      throw new Error('Token is invalid.');
    }
    this.store.update('auth', state => ({
      ...state,
      accessToken: token,
    }));
  }

  removeToken(): void {
    this.store.update('auth', state => ({
      ...state,
      accessToken: undefined,
    }));
  }

  async getOrganization(label: string): Promise<Organization> {
    return this.Organization.get(label);
  }

  async listOrganizations(
    listOrgOptions?: ListOrgOptions,
  ): Promise<PaginatedList<Organization>> {
    try {
      return this.Organization.list(listOrgOptions);
    } catch (error) {
      throw error;
    }
  }

  async createOrganization(
    label: string,
    orgPayload?: CreateOrgPayload,
  ): Promise<Organization> {
    try {
      return this.Organization.create(label, orgPayload);
    } catch (error) {
      throw error;
    }
  }

  async updateOrganization(
    label: string,
    rev: number,
    orgPayload: CreateOrgPayload,
  ): Promise<Organization> {
    try {
      return this.Organization.update(label, rev, orgPayload);
    } catch (error) {
      throw error;
    }
  }

  async deprecateOrganization(
    label: string,
    rev: number,
  ): Promise<Organization> {
    try {
      return this.Organization.deprecate(label, rev);
    } catch (error) {
      throw error;
    }
  }
}
