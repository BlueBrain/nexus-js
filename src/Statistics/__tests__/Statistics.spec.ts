import { GlobalWithFetchMock } from 'jest-fetch-mock';
import Nexus from '../../Nexus';
import Statistics from '../index';
import { getViewStatistics, getResourceStatistics } from '../utils';

const { fetchMock } = <GlobalWithFetchMock>global;
const baseUrl = 'http://api.url';
Nexus.setEnvironment(baseUrl);

const mockStatistics = {
  '@context': 'https://bluebrain.github.io/nexus/contexts/view.json',
  delayInSeconds: 0,
  discardedEvents: 0,
  evaluatedEvents: 10,
  lastEventDateTime: '2019-04-03T12:27:41.344Z',
  lastProcessedEventDateTime: '2019-04-03T12:27:41.344Z',
  processedEvents: 10,
  remainingEvents: 0,
  totalEvents: 10,
};

describe('Statistics', () => {
  describe('getViewStatistics()', () => {
    beforeEach(() => {
      fetchMock.mockResponse(JSON.stringify(mockStatistics), { status: 200 });
    });
    afterEach(() => {
      fetchMock.resetMocks();
    });

    it('should call the appropriate URL', async () => {
      await getViewStatistics('myorg', 'myproject', 'viewid');
      expect(fetchMock.mock.calls[0][0]).toEqual(
        `${baseUrl}/views/myorg/myproject/viewid/statistics`,
      );
    });
    it('should encode the viewID', async () => {
      await getViewStatistics('myorg', 'myproject', 'http://viewid.com');
      expect(fetchMock.mock.calls[0][0]).toEqual(
        `${baseUrl}/views/myorg/myproject/http%3A%2F%2Fviewid.com/statistics`,
      );
    });
    it('should return an instance of statistics', async () => {
      const s: Statistics = await getViewStatistics(
        'myorg',
        'myproject',
        'viewid',
      );
      expect(s).toBeInstanceOf(Statistics);
      expect(s).toMatchSnapshot();
    });
  });
  describe('getResourceStatistics()', () => {
    beforeEach(() => {
      fetchMock.mockResponse(JSON.stringify(mockStatistics), { status: 200 });
    });
    afterEach(() => {
      fetchMock.resetMocks();
    });

    it('should call the appropriate URL', async () => {
      await getResourceStatistics('myorg', 'myproject', 'schemaid', 'viewid');
      expect(fetchMock.mock.calls[0][0]).toEqual(
        `${baseUrl}/resources/myorg/myproject/schemaid/viewid/statistics`,
      );
    });
    it('should encode the viewID', async () => {
      await getResourceStatistics(
        'myorg',
        'myproject',
        'http://schemaid.com',
        'http://viewid.com',
      );
      expect(fetchMock.mock.calls[0][0]).toEqual(
        `${baseUrl}/resources/myorg/myproject/http%3A%2F%2Fschemaid.com/http%3A%2F%2Fviewid.com/statistics`,
      );
    });
    it('should return an instance of statistics', async () => {
      const s: Statistics = await getResourceStatistics(
        'myorg',
        'myproject',
        '_',
        'viewid',
      );
      expect(s).toBeInstanceOf(Statistics);
      expect(s).toMatchSnapshot();
    });
  });
});
