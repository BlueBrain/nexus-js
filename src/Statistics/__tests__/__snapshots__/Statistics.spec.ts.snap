// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statistics getResourceStatistics() should return an instance of statistics 1`] = `
Statistics {
  "context": "https://bluebrain.github.io/nexus/contexts/view.json",
  "delayInSeconds": 0,
  "discardedEvents": 0,
  "evaluatedEvents": 10,
  "lastEventDateTime": "2019-04-03T12:27:41.344Z",
  "lastProcessedEventDateTime": "2019-04-03T12:27:41.344Z",
  "processedEvents": 10,
  "remainingEvents": 0,
  "totalEvents": 10,
}
`;

exports[`Statistics getViewStatistics() should return an instance of statistics 1`] = `
Statistics {
  "context": "https://bluebrain.github.io/nexus/contexts/view.json",
  "delayInSeconds": 0,
  "discardedEvents": 0,
  "evaluatedEvents": 10,
  "lastEventDateTime": "2019-04-03T12:27:41.344Z",
  "lastProcessedEventDateTime": "2019-04-03T12:27:41.344Z",
  "processedEvents": 10,
  "remainingEvents": 0,
  "totalEvents": 10,
}
`;
